torchvision/csrc/models/densenet.cpp

```diff
@@ -0,0 +1,216 @@
+    push_back(torch::nn::BatchNorm(num_input_features));
+    push_back(modelsimpl::Relu(true));
+    push_back(
+        torch::nn::Conv2d(Options(num_input_features, bn_size * growth_rate, 1)
+                              .stride(1)
+                              .with_bias(false)));
+    push_back(torch::nn::BatchNorm(bn_size * growth_rate));
+    push_back(modelsimpl::Relu(true));
+    push_back(torch::nn::Conv2d(Options(bn_size * growth_rate, growth_rate, 3)
+                                    .stride(1)
+                                    .padding(1)
+                                    .with_bias(false)));
+  }
+
+  torch::Tensor forward(torch::Tensor x) {
+    auto new_features = torch::nn::SequentialImpl::forward(x);
+    if (drop_rate > 0)
+      new_features =
+          torch::dropout(new_features, drop_rate, this->is_training());
+    return torch::cat({x, new_features}, 1);
+  }
+};
+
+TORCH_MODULE(_DenseLayer);
+
+class _DenseBlockImpl : public torch::nn::SequentialImpl {
+ public:
+  _DenseBlockImpl(
+      int64_t num_layers,
+      int64_t num_input_features,
+      int64_t bn_size,
+      int64_t growth_rate,
+      double drop_rate) {
+    for (int64_t i = 0; i < num_layers; ++i) {
+      auto layer = _DenseLayer(
+          num_input_features + i * growth_rate,
+          growth_rate,
+          bn_size,
+          drop_rate);
+      push_back(layer);
+    }
+  }
+
+  torch::Tensor forward(torch::Tensor x) {
+    return torch::nn::SequentialImpl::forward(x);
+  }
+};
+
+TORCH_MODULE(_DenseBlock);
+
+class _TransitionImpl : public torch::nn::SequentialImpl {
+ public:
+  _TransitionImpl(int64_t num_input_features, int64_t num_output_features) {
+    push_back(torch::nn::BatchNorm(num_input_features));
+    push_back(modelsimpl::Relu(true));
+    push_back(
+        torch::nn::Conv2d(Options(num_input_features, num_output_features, 1)
+                              .stride(1)
+                              .with_bias(false)));
+    push_back(modelsimpl::AvgPool2D(2, 2));
+  }
+
+  torch::Tensor forward(torch::Tensor x) {
+    return torch::nn::SequentialImpl::forward(x);
+  }
+};
+
+TORCH_MODULE(_Transition);
+
+DenseNetImpl::DenseNetImpl(
+    int64_t num_classes,
+    int64_t growth_rate,
+    std::vector<int64_t> block_config,
+    int64_t num_init_features,
+    int64_t bn_size,
+    double drop_rate) {
+  // First convolution
+  features = torch::nn::Sequential(
+      torch::nn::Conv2d(torch::nn::Conv2dOptions(3, num_init_features, 7)
+                            .stride(2)
+                            .padding(3)
+                            .with_bias(false)),
+      torch::nn::BatchNorm(num_init_features),
+      modelsimpl::Relu(true),
+      modelsimpl::MaxPool2D(
+          modelsimpl::MaxPool2DOptions(3).stride(2).padding(1)));
+
+  // Each denseblock
+  auto num_features = num_init_features;
+  for (size_t i = 0; i < block_config.size(); ++i) {
+    auto num_layers = block_config[i];
+    auto block =
```

Comment:
It's more C++-y to write `_DenseBlock block(...)`. `auto obj = Class()` is more of a python-ism

Reply:
I did this so much in casting that I got used to it. I'll fix this